Refactor input.Event similar to outputs.Data

vjsamuel opened the pull request:
Kindly review this change as this is a feature that we're looking for when we use filebeat :) If the code looks good, I can also fix the test cases.

https://github.com/elastic/beats/issues/3785

cc: @ruflin

elasticmachine commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run.

elasticmachine commented:
Jenkins standing by to test this. If you aren't a maintainer, you can ignore this comment. Someone with commit access, please review this and clear it for Jenkins to run.

ruflin commented on filebeat/f.yml:
```diff
@@ -0,0 +1,143 @@
+###################### Filebeat Configuration Example #########################
```
that was probably commited by accident. You can use filebeat.dev.yml and it will be ignored automatically by gitignore.

vjsamuel replied:
removed it. thanks for the heads up. i ll use that.

ruflin commented on filebeat/prospector/config.go:
```diff
@@ -43,6 +44,7 @@ type prospectorConfig struct {
 	Pipeline             string             `config:"pipeline"`
 	Module               string             `config:"_module_name"`  // hidden option to set the module name
 	Fileset              string             `config:"_fileset_name"` // hidden option to set the fileset name
+	Filters              processors.PluginConfig `config:"filters"`
```
Probably we should call it processors, as all of them will be available. It's already causing confusion on the metricbeat side ...

vjsamuel replied:
ha ha! let me do that.

ruflin commented on filebeat/prospector/prospector.go:
```diff
@@ -214,12 +223,20 @@ func (p *Prospector) updateState(event *input.Event) error {
+
+	//run the filters before sending to
+	eventMap = p.filters.Run(eventMap)
+	if eventMap != nil {
```
Assuming the event was "removed" by the filter, this would probably mean no state update is sent anymore? Or when exactly is `eventMap` empty?

vjsamuel replied:
eventMap would be empty in case of `drop_event`. But i did see that the file was traversed till the end:

```
2017/03/28 08:17:17.442705 log.go:114: INFO Reader was closed: /var/log/alf.log. Closing.
2017/03/28 08:17:17.443151 log.go:306: DBG  Stopping harvester for file: /var/log/alf.log
2017/03/28 08:17:17.443163 log.go:315: DBG  Closing file: /var/log/alf.log
2017/03/28 08:17:17.443169 log.go:183: DBG  Update state: /var/log/alf.log, offset: 0
2017/03/28 08:17:17.443130 log.go:183: DBG  Update state: /var/log/displaypolicyd.stdout.log, offset: 1155
2017/03/28 08:17:17.443030 log.go:315: DBG  Closing file: /var/log/jamf.log
2017/03/28 08:17:17.443198 log.go:183: DBG  Update state: /var/log/jamf.log, offset: 1190762
```

But since these events do goto registrar, the registry file wouldn't be updated. But this is a scenario where I configure my prospector with:
```
filters:
  drop_event:
```

In this case, no event would pass upto registrar.

ruflin commented:
jenkins, test it

ruflin commented on filebeat/input/event.go:
```diff
@@ -11,24 +11,37 @@ import (
+	State    file.State
+}
+
+type EventHolder struct {
```
Can we rename `EventHolder` to `Data` to be consistent with https://github.com/elastic/beats/blob/master/libbeat/outputs/outputs.go#L20 ? I'm ok with keeping Metadata as here it is at the moment only about Metadata

ruflin commented on filebeat/input/event.go:
```diff
@@ -11,24 +11,37 @@ import (
 	Module       string
+	InputType    string
+	DocumentType string
+	common.EventMetadata
```
lets move the common.EventMetadata to the the first line of EventMeta

vjsamuel replied:
done.

ruflin commented on filebeat/input/event.go:
```diff
@@ -11,24 +11,37 @@ import (
-	JSONConfig   *reader.JSONConfig
-	State        file.State
-	Data         common.MapStr // Use in readers to add data to the event
+	EventMeta
```
So currently we have EventMeta twice, means two copies of the data? What would be the affect if we not have it in the Event? I think we will need in the future processors to have access to the meta data too, but do we need this in the first step?

vjsamuel replied:
Event will be used in harvesters and when it reaches the processor, we do a event.GetEventHolder(), which would (you are right), copy the metadata from Event to EventHolder and generate the MapStr. But after that since Event object goes out of scope, it would be garbage collected if I am not wrong. Would that be an issue? Today when we generate the common.MapStr, we copy all the log related fields over. Now with this change, we additionally copy over the meta fields as well. Please correct me if I'm wrong.

ruflin commented on filebeat/prospector/prospector.go:
```diff
@@ -214,7 +223,22 @@ func (p *Prospector) updateState(event *input.Event) error {
+	var ok bool
+	if eventHolder.Event != nil {
+		//processor might decide to drop the event
+		ok = p.outlet.OnEvent(&eventHolder)
```
Lets move the `ok = p.outlet.OnEvent(&eventHolder)` after the if / else and only have `if eventHolder.Event == nil { Bytes = 0} `. This should be the same logic. Then we also don't need the definition of ok.

vjsamuel replied:
will add this back in the next PR. removed it for now.

vjsamuel commented:
@ruflin, i have removed the processor code and renamed the object from `EventHolder` to `Data`.

Also do I submit the PR for processor code immediately or do I wait for this to be approved and merged?

ruflin commented:
jenkins, test it

vjsamuel commented:
thank you! :)
